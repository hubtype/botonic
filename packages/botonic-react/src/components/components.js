/**
 * Get the name of an element (instantiated component)
 * @param {React.Element} component
 * @returns {string}
 */
export function getElementName(component) {
  return component.type.name
}

/**
 * Get the name of a component TYPE
 * @param {React.ComponentType} component
 * @returns {string}
 */
export function getComponentTypeName(component) {
  return component.displayName || component.name || 'Component'
}
