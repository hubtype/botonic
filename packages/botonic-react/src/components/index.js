export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export { Location } from './location'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export { WhatsappTemplate } from './whatsapp-template'
