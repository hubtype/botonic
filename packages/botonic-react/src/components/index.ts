export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export * from './index-types'
export { Location } from './location'
export { Message } from './message'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Raw } from './raw'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export {
  WhatsappButtonList,
  WhatsappButtonListProps,
  WhatsappButtonListRowProps,
  WhatsappButtonListSectionProps,
} from './whatsapp-button-list'
export {
  WhatsappCTAUrlButton,
  WhatsappCTAUrlButtonProps,
} from './whatsapp-cta-url-button'
export { WhatsappTemplate } from './whatsapp-template'
