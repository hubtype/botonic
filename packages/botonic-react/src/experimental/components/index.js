export { Button } from '../../components/button'
export { Element } from '../../components/element'
export { MessageTemplate } from '../../components/message-template'
export * from '../../components/multichannel'
export { Pic } from '../../components/pic'
export { Reply } from '../../components/reply'
export { ShareButton } from '../../components/share-button'
export { Subtitle } from '../../components/subtitle'
export { Title } from '../../components/title'
export { WebchatSettings } from '../../components/webchat-settings'

// Experimental
export { Audio } from './audio'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Image } from './image'
export { Location } from './location'
export { Text } from './text'
export { Video } from './video'
export { WhatsappTemplate } from './whatsapp-template'
