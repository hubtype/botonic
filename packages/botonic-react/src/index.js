export { BotonicInputTester, BotonicOutputTester } from './botonic-tester'
export * from './components'
export { RequestContext, WebchatContext } from './contexts'
export { DevApp } from './dev-app'
export { msgsToBotonic, msgToBotonic } from './msg-to-botonic'
export { NodeApp } from './node-app'
export { staticAsset } from './util/environment'
export { getBotonicApp, Webchat } from './webchat/index.js'
export { WebchatApp } from './webchat-app'
export { WebviewApp } from './webview'
