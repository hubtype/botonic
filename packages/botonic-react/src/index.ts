const React = require('react')
const ReactDOMServer = require('react-dom/server')

import { default as i18n } from '@botonic/core/lib/i18n'

declare global {
    namespace JSX {
        interface IntrinsicElements {
            [elemName: string]: any;
        }
    }
}

export async function renderReactAction(req, action) {
	let props = {}
	if(action.botonicInit) {
       props = await action.botonicInit({req})
    }

    let component = React.createElement(action, (props && typeof(props)==='object') ? Object.assign(props, req) : req)
    return ReactDOMServer.renderToStaticMarkup(component)
}

export function setLocale(req: any, locale: string) {
    req.context.__locale = locale
    i18n.setLocale(locale)
}
export { BotonicWebview } from './webview'
export { MessageTemplate } from './messageTemplate'
export { ShareButton } from './shareButton'
