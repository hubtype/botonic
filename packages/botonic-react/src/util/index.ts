export * from './dom'
export * from './environment'
export * from './error-boundary'
export * from './logs'
export * from './objects'
export * from './react'
export * from './regexs'
export * from './strings'
export * from './webchat'
