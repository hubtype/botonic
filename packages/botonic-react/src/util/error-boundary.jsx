import { isNode } from '@botonic/core'
import React from 'react'

import { Text } from '../components/text'

/**
 * Replaces crashed children with the provided fallback component.
 * https://reactjs.org/blog/2017/07/26/error-handling-in-react-16.html
 * See alternative at https://stackoverflow.com/a/60255291/145289
 */
export const createErrorBoundary = ({
  errorComponent = props => <Text>The message cannot be displayed</Text>,
} = {}) => {
  class ErrorBoundary extends React.Component {
    constructor(props) {
      super(props)
      this.state = { error: null }
    }

    /**
     * @param error the exception which was trown
     * @param errorInfo the stack of component names at the error
     */
    componentDidCatch(error, errorInfo) {
      // No need to log the error because at least chrome & firefox already show
      // both component and call stacks
      if (isNode()) {
        // In node, only the component stack is displayed
        console.error(`Failure at:`, error)
      }
    }

    static getDerivedStateFromError(error) {
      return { error }
    }

    render() {
      if (this.state.error) {
        return errorComponent({
          ...this.props,
          errorMessage: this.state.error.message,
        })
      } else {
        return this.props.children
      }
    }
  }

  return ErrorBoundary
}
