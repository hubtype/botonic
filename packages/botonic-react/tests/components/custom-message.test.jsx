import React from 'react'
import TestRenderer from 'react-test-renderer'

import { Button, customMessage, Reply, Text } from '../../src/components'
import { getElementName } from '../../src/components/components'
import { createErrorBoundary } from '../../src/util/error-boundary'

const renderToJSON = sut => TestRenderer.create(sut).toJSON()

test('TEST: CustomMessage defaultProps, props and children are injected into the component', () => {
  class CustomComponent extends React.Component {
    render() {
      return <Text>{this.props.wrappedProp + this.props.children}</Text>
    }
  }

  const Sut = customMessage({
    name: 'test1',
    component: CustomComponent,
    defaultProps: { wrapperProp: 'wrapperPropV' },
  })
  const tree = renderToJSON(<Sut wrappedProp='wrappedPropV'>body</Sut>)
  expect(tree).toMatchSnapshot()
})

describe('CustomMessage replies', () => {
  function names(children) {
    if (Array.isArray(children)) {
      return children.map(c => getElementName(c)).join(',')
    }
    return getElementName(children)
  }

  function reply(key) {
    return (
      <Reply key={key} payload={`payload${key}`}>
        reply{key}
      </Reply>
    )
  }

  function button(key) {
    return (
      <Button key={key} payload={`payload${key}`}>
        button{key}
      </Button>
    )
  }

  for (const children of [
    reply(), // single reply
    button(), // single no reply
    [reply(1), reply(2)], // multiple all reply
    [reply(1), button(2)], // mix reply & no reply
    [button(1), button(2)], // multiple no reply
  ]) {
    test(`TEST: replies are moved outside the custom component ${names(
      children
    )}`, () => {
      class CustomComponent extends React.Component {
        render() {
          return <Text>{this.props.children}</Text>
        }
      }

      const Sut = customMessage({
        name: 'test1',
        component: CustomComponent,
      })
      const tree = renderToJSON(<Sut>{children}</Sut>)
      expect(tree).toMatchSnapshot()
    })
  }
})

describe('CustomMessage failing', () => {
  class FailingComponent extends React.Component {
    // eslint-disable-next-line react/require-render-return
    render() {
      throw Error('Forced exception')
    }
  }

  test('TEST: ErrorBoundary on a failing component', () => {
    const Sut = customMessage({
      name: 'test1',
      component: FailingComponent,
      defaultProps: { wrapperProp: 'wrapperPropV' },
    })
    const tree = renderToJSON(<Sut />)
    expect(tree).toMatchSnapshot()
  })

  test('TEST: ErrorBoundary different errorComponent which gets the props', () => {
    const Sut = customMessage({
      name: 'test1',
      component: FailingComponent,
      errorBoundary: createErrorBoundary({
        // eslint-disable-next-line react/display-name
        errorComponent: props => (
          <Text>{`Boom!${props.bang}: ${props.errorMessage}`}</Text>
        ),
      }),
    })
    const tree = renderToJSON(<Sut bang='Bang!' />)
    expect(tree).toMatchSnapshot()
  })
})
