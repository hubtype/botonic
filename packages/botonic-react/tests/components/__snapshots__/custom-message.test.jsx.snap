// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomMessage failing TEST: ErrorBoundary different errorComponent which gets the props 1`] = `
<message
  delay={0}
  json="{\\"bang\\":\\"Bang!\\",\\"customTypeName\\":\\"test1\\"}"
  type="custom"
  typing={0}
>
  <message
    delay={0}
    markdown={1}
    type="text"
    typing={0}
  >
    Boom!Bang!: Forced exception
  </message>
</message>
`;

exports[`CustomMessage failing TEST: ErrorBoundary on a failing component 1`] = `
<message
  delay={0}
  json="{\\"customTypeName\\":\\"test1\\"}"
  type="custom"
  typing={0}
  wrapperProp="wrapperPropV"
>
  <message
    delay={0}
    markdown={1}
    type="text"
    typing={0}
  >
    The message cannot be displayed
  </message>
</message>
`;

exports[`CustomMessage replies TEST: replies are moved outside the custom component Button 1`] = `
<message
  delay={0}
  json="{\\"children\\":{\\"key\\":null,\\"ref\\":null,\\"props\\":{\\"payload\\":\\"payloadundefined\\",\\"children\\":[\\"button\\",null]},\\"_owner\\":null,\\"_store\\":{}},\\"customTypeName\\":\\"test1\\"}"
  type="custom"
  typing={0}
>
  <message
    delay={0}
    markdown={1}
    type="text"
    typing={0}
  >
    <button
      payload="payloadundefined"
    >
      button
    </button>
  </message>
</message>
`;

exports[`CustomMessage replies TEST: replies are moved outside the custom component Button,Button 1`] = `
<message
  delay={0}
  json="{\\"children\\":[{\\"key\\":\\".$1\\",\\"ref\\":null,\\"props\\":{\\"payload\\":\\"payload1\\",\\"children\\":[\\"button\\",1]},\\"_owner\\":null,\\"_store\\":{}},{\\"key\\":\\".$2\\",\\"ref\\":null,\\"props\\":{\\"payload\\":\\"payload2\\",\\"children\\":[\\"button\\",2]},\\"_owner\\":null,\\"_store\\":{}}],\\"customTypeName\\":\\"test1\\"}"
  type="custom"
  typing={0}
>
  <message
    delay={0}
    markdown={1}
    type="text"
    typing={0}
  >
    <button
      payload="payload1"
    >
      button
      1
    </button>
    <button
      payload="payload2"
    >
      button
      2
    </button>
  </message>
</message>
`;

exports[`CustomMessage replies TEST: replies are moved outside the custom component Reply 1`] = `
<message
  delay={0}
  json="{\\"children\\":[],\\"customTypeName\\":\\"test1\\"}"
  type="custom"
  typing={0}
>
  <message
    delay={0}
    markdown={1}
    type="text"
    typing={0}
  />
  <reply
    payload="payloadundefined"
  >
    reply
  </reply>
</message>
`;

exports[`CustomMessage replies TEST: replies are moved outside the custom component Reply,Button 1`] = `
<message
  delay={0}
  json="{\\"children\\":[{\\"key\\":\\".$2\\",\\"ref\\":null,\\"props\\":{\\"payload\\":\\"payload2\\",\\"children\\":[\\"button\\",2]},\\"_owner\\":null,\\"_store\\":{}}],\\"customTypeName\\":\\"test1\\"}"
  type="custom"
  typing={0}
>
  <message
    delay={0}
    markdown={1}
    type="text"
    typing={0}
  >
    <button
      payload="payload2"
    >
      button
      2
    </button>
  </message>
  <reply
    payload="payload1"
  >
    reply
    1
  </reply>
</message>
`;

exports[`CustomMessage replies TEST: replies are moved outside the custom component Reply,Reply 1`] = `
<message
  delay={0}
  json="{\\"children\\":[],\\"customTypeName\\":\\"test1\\"}"
  type="custom"
  typing={0}
>
  <message
    delay={0}
    markdown={1}
    type="text"
    typing={0}
  />
  <reply
    payload="payload1"
  >
    reply
    1
  </reply>
  <reply
    payload="payload2"
  >
    reply
    2
  </reply>
</message>
`;

exports[`TEST: CustomMessage defaultProps, props and children are injected into the component 1`] = `
<message
  delay={0}
  json="{\\"wrappedProp\\":\\"wrappedPropV\\",\\"children\\":\\"body\\",\\"customTypeName\\":\\"test1\\"}"
  type="custom"
  typing={0}
  wrapperProp="wrapperPropV"
>
  <message
    delay={0}
    markdown={1}
    type="text"
    typing={0}
  >
    wrappedPropVbody
  </message>
</message>
`;
